tests/ui/rfcs/rfc-0000-never_patterns/diverges-not.rs

```diff
@@ -15,7 +15,6 @@ fn wild_void(_: Void) -> u32 {}
 fn wild_let() -> u32 {
     let ptr: *const Void = std::ptr::null();
     unsafe {
-        //~^ ERROR: mismatched types
         let _ = *ptr;
     }
 }
```

Comment:
This is incorrect: we should *not* infer that this function diverges, because `let _ = *ptr;` does not cause a read. In fact `let _ = *std::ptr::null::<Void>();` is entirely value code that creates a null pointer and does nothing with it. Typechecking should distinguish place expressions from value expressions in order to be correct.

Reply:
Is this not pre-existing though? I did not have time to work on https://github.com/rust-lang/rust/issues/117288 and neither anyone else as far as I'm aware. Do we have any infrastructure to distinguish places & values in hir?

Reply:
Idk if that's the same issue, but we shouldn't regress behavior.

Place vs value is just a matter of matching on ExprKind, there's probably a helper somewhere to distinguish the two

Reply:
So what you are saying is that an expression is a place iff it's `Unary(Deref, _)`, `Index(..)` or `Field(..)`? But are those always places? Or does it not matter because in cases where we care they are places?...

Reply:
Please use `MemCategorizationContext`, I think that's the right helper for this

Reply:
There is https://github.com/rust-lang/rust/blob/bb78dba64ca4158ef2f3488d0d41a82c75a504f2/compiler/rustc_hir/src/hir.rs#L1622-L1626 but yeah, `MemCategorizationContext` looks right for this, thanks for the pointer!